core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -407,30 +419,106 @@ public void index(Index index) {
         }
     }
 
+    private boolean canOptimizeAddDocument(Index index) {
```

Comment:
I made all of the asserts I think we should still keep it as it contains it well